Adapt tests to support CI and broad them to other projects

amurKontur opened the pull request:
Expand login tests to cover all key projects and fix config

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
    - Updated the `.env.playwright` file to include project details for improved test automation.
    - Added functionality to dynamically generate login tests for various projects using Playwright.
- **Documentation**
    - Enhanced guidelines for disaster-ninja front-end build with Docker and Vite.
    - Updated i18n instructions and translation steps in React.
    - Provided detailed information on tests, type checks, and configuration updates.
- **Refactor**
    - Restructured color mapping logic in `mcdaStyle.ts` for better rendering.
    - Improved logic for determining cell labels in `bivariateLegendUtils.ts`.
    - Enhanced type safety and function clarity in `styleGen.ts`.
- **Tests**
    - Added edge case tests for `getCellLabelByValue` function in `bivariateLegendUtils.test.ts`.
- **Chores**
    - Updated exclusions in coverage configuration in `vite.config.ts`.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
Expand login tests to cover all key projects and fix config

Commit:
Expand login tests to cover all key projects and fix config

dqunbp committed:
fix: rename tests examples (#682)

dqunbp commented on .env.playwright:
It is better to keep this file as is; it contains local environment configurations which will be overridden by the `.env.playwright.local` file.

amurKontur replied:
but it now supports json, not just url

dqunbp commented on .env.playwright:
```diff
@@ -1 +1,23 @@
-BASE_URL=https://localhost:3000
\ No newline at end of file
+PROJECTS='{
```
It looks like it will be more maintainable to save this JSON config to a JSON file and then read from it when the tests start.

amurKontur replied:
I tried to do it, but it seems like some issues appear than. We should better postpone until we know how to make a workable beautiful playwright config

dqunbp replied:
What kind of issues appeared?

amurKontur replied:
fixed

amurKontur commented on .env.playwright:
```diff
@@ -1 +1,23 @@
+        "app":"1dc6fe68-8802-4672-868d-7f17943bf1c8",
+        "title":"OpenAerialMap"
+    }
+}'
```
app IDs are available publicly at our websites urls' params

amurKontur commented on e2e/login.spec.ts:
```diff
@@ -0,0 +1,53 @@
+const projects = JSON.parse(process.env.PROJECTS!) as Record<
+  string,
+  { url: string; app: string; title: string }
+>;
```
project variable is a global one, without it no testing is possible

dqunbp commented on README.md:
````diff
@@ -182,7 +208,29 @@ To view the app, visit https://localhost:3000 in your browser.
 Create a `.env.playwright.local` file in the root of the project with the following content:
 
 ```bash
-BASE_URL=app_base_url
````
It looks like we use absolute app URLs; it's not a base URL anymore.

If we will not use the base URL, just remove it completely.

amurKontur replied:
deleted

amurKontur commented on e2e/login.spec.ts:
```diff
@@ -0,0 +1,53 @@
+    // Currently OAM and Atlas apps don't have cookies popups
+    if (titles[i] !== projects.oam.title && titles[i] !== projects.atlas.title) {
+      await page.getByText('Accept optional cookies').click();
+    }
```
Here the bug is present that was found by autotest - https://kontur.fibery.io/Tasks/Task/Add-cookie-feature-to-Kontur-apps-18184. After its fix, if clause should be deleted. Just click on popup should stay

dqunbp replied:
Please follow the coderabbit))

amurKontur replied:
@coderabbitai please open an issue

amurKontur commented on playwright.config.ts:
```diff
@@ -25,7 +25,7 @@ export default defineConfig({
   use: {
     /* Base URL to use in actions like `await page.goto('/')`. */
-    baseURL: process.env.BASE_URL,
+    baseURL: JSON.parse(process.env.PROJECTS!).atlas.url,
```
no, our main focus is Atlas, so we should use it as base url with majority of tests focusing on this project

amurKontur commented on README.md:
````diff
@@ -32,33 +35,39 @@ pnpm run serve
 ### Scripts
+
 We use i18next + gettext convertion approach in our development toolchain.
 There are several scripts in package.json:
 
 - i18n:export converts i18next to gettext
 - i18n:import converts gettext to i18next
-- i18n:update extracts new translation keys from code to common.json (```i18n.t``` or ```<Trans/>```) and also removes unused keys
+- i18n:update extracts new translation keys from code to common.json (`i18n.t` or `<Trans/>`) and also removes unused keys
 - lint:i18n:keys:identity is needed to see untranslated keys for all locales (make sure you ran i18n:import before as it compares i18next .json files)
 - i18n:gettext-sync is needed to keep all .po files synced with .pot file (marks non-compliant msgid as 'fuzzy', renames keys in .po according keys changes in .pot)
````
this changes were done by prettier, not sure that we need to keep them, better get some other dev opinion

dqunbp commented on e2e/login.spec.ts:
```diff
@@ -0,0 +1,53 @@
+
+// Create old-school loop to loop over all the projects and create a test for everyone
+
+for (let i = 0; i < urls.length; i++) {
```
Please iterate over the projects itself, for example `for (project of projects)`

amurKontur replied:
done

dqunbp commented on e2e/login.spec.ts:
```diff
@@ -0,0 +1,53 @@
+
+// Create old-school loop to loop over all the projects and create a test for everyone
+
+for (let i = 0; i < urls.length; i++) {
```
Please iterate over the projects itself, for example `for (project of projects)`

amurKontur replied:
done

IlyaLosikKontur committed:
fix: 18065: correct bivariate legend style calculation rules (#680)

* fix: 18065: correct bivariate legend style calculation rules

* fix: 18065: add tests and types

albaranau committed:
fix(18075): don't show out-of range MCDA results as transparent for "sentiments" color config (#684)

Commit:
added simplified loop and removed base url

Commit:
fix conflicts

dqunbp commented on e2e/login.spec.ts:
```diff
@@ -0,0 +1,45 @@
+// Create a loop to loop over all the projects and create a test for everyone
+
+for (const project of Object.keys(projects)) {
+  test(`As User, I can login to ${projects[project].title}`, async ({ page }) => {
```
We can do it without `Object.keys` then accessing to the properties directly just `project.title`, etc

amurKontur replied:
better way was used finally

Commit:
fixed projects, moved projects configs to separate json, fixed logics of atlas check

Commit:
fix readme

Commit:
moved projects-config to e2e folder

Commit:
Added CI pipeline

Commit:
fix pipeline

Commit:
fix pipeline name

Commit:
fixed pipeline name

Commit:
fix rabbit review

amurKontur commented on .github/workflows/qa.yml:
```diff
@@ -0,0 +1,80 @@
+          fi
+
+  - name: Move to e2e directory # Move to e2e
+    run: cd e2e
```
done

Commit:
fix error of pipeline steps

Commit:
added shell call

Commit:
Added timeout and node.js config installation

Commit:
Moved input to separate variable as for rabbit suggestion

Commit:
fix syntax

amurKontur commented on .github/workflows/run_e2e_tests.yml:
```diff
@@ -0,0 +1,70 @@
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$dev_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$dev_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$dev_oam_url\"|g" e2e/projects-config.json
+          fi
```
fixed

amurKontur commented on .github/workflows/run_e2e_tests.yml:
```diff
@@ -0,0 +1,77 @@
+          elif [[ "${{ github.event.inputs.env }}" == "prod" ]]; then
+            echo "TEST_EMAIL=${{ secrets.PROD_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.PROD_PASSWORD }}" >> .env.playwright.local
+          fi
```
fixed

dqunbp commented on .github/workflows/run_e2e_tests.yml:
```diff
@@ -0,0 +1,79 @@
+        uses: actions/setup-node@v3
+        with:
+          node-version: '20'
+      - name: Install dependencies
```
We don't need the project dependencies

amurKontur commented on .github/workflows/e2e_tests.yml:
```diff
@@ -0,0 +1,71 @@
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$dev_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$dev_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$dev_oam_url\"|g" e2e/projects-config.json
+          fi
```
fixed in new version

amurKontur commented on .github/workflows/e2e_tests.yml:
```diff
@@ -0,0 +1,71 @@
+          elif [[ "${{ github.event.inputs.env }}" == "prod" ]]; then
+            echo "TEST_EMAIL=${{ secrets.PROD_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.PROD_PASSWORD }}" >> .env.playwright.local
+          fi
```
fixed in new version

amurKontur commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,80 @@
+          fi
+
+      - name: Move to e2e directory # Move to e2e
+        run: cd e2e
```
fixed in new version

amurKontur commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,80 @@
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$dev_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$dev_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$dev_oam_url\"|g" e2e/projects-config.json
+          fi
```
fixed in new version

amurKontur commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,80 @@
+          elif [[ "${{ github.event.inputs.env }}" == "prod" ]]; then
+            echo "TEST_EMAIL=${{ secrets.PROD_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.PROD_PASSWORD }}" >> .env.playwright.local
+          fi
```
fixed in new version

amurKontur commented on .github/workflows/qa.yml:
```diff
@@ -0,0 +1,80 @@
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$dev_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$dev_oam_url\"|g" e2e/projects-config.json
+          fi
+
```
fixed in new version

amurKontur commented on .github/workflows/qa.yml:
```diff
@@ -0,0 +1,80 @@
+          elif [[ "${{ github.event.inputs.env }}" == "prod" ]]; then
+            echo "TEST_EMAIL=${{ secrets.PROD_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.PROD_PASSWORD }}" >> .env.playwright.local
+          fi
```
fixed in new version

amurKontur commented on .github/workflows/qa.yml:
```diff
@@ -0,0 +1,80 @@
+
+      - name: Move to e2e directory # Move to e2e
+        run: cd e2e
+
```
fixed in new version

dqunbp committed:
Refactor e2e test configuration and environment setup (#692)

This commit introduces several changes to the end-to-end (e2e) testing setup and configuration. The primary changes include:

- Standardization of environment variable names across different files, replacing `TEST_EMAIL` and `TEST_PASSWORD` with `EMAIL` and `PASSWORD`.
- Removal of hardcoded environment URLs in the GitHub Actions workflow file, replacing them with a dynamic setup that adjusts based on the input environment.
- Deletion of an example test file that was no longer needed.
- Update of the login test specification to accommodate changes in the environment setup and to remove hardcoded values.
- Refactoring of the projects configuration JSON to include environment-specific URLs and additional properties such as `hasCookieBanner`.
- Adjustment of the Playwright configuration to prioritize the production environment file.

These changes aim to simplify the configuration process, reduce duplication, and make the e2e testing framework more flexible and maintainable.
